BASEX improvements

MikhailRyazanov opened the pull request:
This addresses all issues in #230 and finalizes #226.

## Main changes

The basis-generation code (`_bs_basex()`) is rewritten almost completely.
- Now it works with arbitrary width parameters _σ_ and accepts `n, sigma` instead of `n, nbf`. `sigma` is actually used everywhere, since it is more fundamental than `nbf` (which is `n` / `sigma`).
- The computations became is >10 times faster:
   - All the needed Γ-function values are precomputed, which takes _O_(_n_<sup>2</sup>) extra space (~the same as for the basis matrices), but avoids recomputing them _O_(_n_<sup>3</sup>) times.
   - The summation cutoff is tightened very close to the limit for "numerically exact" results (floating-point precision).
- It can take previously computed smaller basis matrices for the same _σ_ and extend them by computing only the missing (outer) elements.

The `get_bs_cached()` code (renamed from `get_bs_basex_cached()` for consistency) is changed accordingly.
- The file-naming scheme is changed to `basex_basis_{n}_{sigma}.npy`.
- Attempts to get the basis (if not already in memory) are tried in the following order:
   1. If the file for given `n`, `sigma` exists, then it is used.
   2. If files for larger `n` and the same `sigma` exist, then the smallest of them is loaded and cropped.
   3. If files for smaller `n` and the same `sigma` exist, then the largest of them is loaded and extended by `_bs_basex()`.
   4. If nothing helps, then a completely new basis is generated.

The `MAGIC_NUMBER` intensity correction is replaced by `get_basex_correction()`, activated by passing `correction=True` to `basex_transform()` (or `get_bs_cached()`). See #230 for details.

A new module `abel.tools.polynomial` is added (primarily for `get_basex_correction()`), which implements the analytical Abel transform for arbitrary polynomials. `Polynomial` and `PiecewisePolynomial` are added to `abel.tools.analytical`.

Test are added/updated for `abel.basex` and `abel.tools.polynomial`.

## Additional changes

Declared `_source` as `global` in `dash.cache_cleanup()` (amending #232).

Minor corrections to documentation.
In particular, `abel.tools.analytical` docstrings are moved from `__init__()` to class levels, since they were not visible in the generated documentation.

## Questions

What is `mask_valid` in `BaseAnalytical` and how to set it meaningfully for polynomials?

What happened to the `GaussianAnalytical` docstring (the [`ratio_valid_sigma` description](https://github.com/PyAbel/PyAbel/blob/be2d1adb1b20a664ac6f757d1c40fb737222ba63/abel/tools/analytical.py#L187-L190))?

The are some problems with multiple parameters in docstrings, for example, see [`abel.tools.polar.cart2polar`](https://pyabel.readthedocs.io/en/latest/abel.html#abel.tools.polar.cart2polar) — `x, y : floats or arrays` is rendered as "**y** (_x,_)".
This is probably related to https://github.com/sphinx-doc/sphinx/issues/2227?

Perhaps, `linbasex.basis_cache_cleanup()` should be renamed to `linbasex.cache_cleanup()` for consistency with and `dasch.cache_cleanup()` `basex.cache_cleanup()`?

MikhailRyazanov commented on doc/abel.rst:
```diff
@@ -169,3 +177,13 @@ abel.benchmark module
     :undoc-members:
     :show-inheritance:
+
+abel.tests module
```
I think, @rth has removed this in 5636c402c3791424859335f8d78ddb41ac47e5e9. I don't know how it got back. Need to remove again?

rth replied:
Yes, there is no point in including tests as part of the public API in the docs.

MikhailRyazanov replied:
There is also `test_suite="abel.tests.run_cli"` remaining [in setup.py](https://github.com/PyAbel/PyAbel/blob/be2d1adb1b20a664ac6f757d1c40fb737222ba63/setup.py#L116). Perhaps, should be removed or updated?

MikhailRyazanov replied:
OK, removed this and `test_suite="abel.tests.run_cli"`.

MikhailRyazanov commented:
_Who will test the testers?_

I am wondering why AppVeyor tests with Python 3.5 fail.

Most interestingly, the 32- and 64-bit builds fail at different checks in `test_basex_basis_sets_resize()`, all of which pass without problems in Travis CI and on both my computers (i5 GNU/Linux and Ryzen3 Win10). And they could not fail the way the do — producing different numbers in the arrays — since the only difference is which of these numbers are calculated and which are copied from the previous calculations with the same parameters.

It seems that their machines have some memory corruption or slightly broken floating-point units. Is it possible to rerun these tests? Or if somebody can reproduce the problem...

stggh commented:
The default, `correction=False`. Oh, you are right ~1.5%, I was looking at profile 2, but its maximum is at the origin. 1.5% correct then?

MikhailRyazanov commented:
It should be (2/√π) / `MAGIC_NUMBER` ≈ 1.0145, that is, ~1.45%.

Try `correction=True`. It should give proper normalization and remove the glitch near _r_ = 0.

Also using `sigma` ≠ 1 (with default `correction=False`) should produce oscillating but averagely correct intensities. Then `correction=True` should also remove these oscillations.

stggh commented:
Yes, that's it. Perhaps, `correction=True` should be the default?

MikhailRyazanov commented:
I also think so, but wanted to make sure that the fundamental stuff works properly and then that this new correction does not cause problems.

DanHickstein commented:
This is fantastic! Well done @MikhailRyazanov. 

> I have changed the assert np.array_equal() test to assert_allclose() with 10−15 tolerances to avoid this problem.

Yeah, `allclose()` is way nicer than checking for perfect equality. It's amazing how often Python thinks that two arrays that should be equal aren't actually equal :)

> What is mask_valid in BaseAnalytical and how to set it meaningfully for polynomials?

I think that @rth could give you the best answer to this question. But, I think that it's an array that is the same length as the returned array and serves as a True/False mask for where the transform is valid. I think that some functions only had a certain range where they were correct.

> What happened to the GaussianAnalytical docstring (the ratio_valid_sigma description)?

That docstring appears to be messed up. Maybe open an Issue?

> The are some problems with multiple parameters in docstrings, for example, see abel.tools.polar.cart2polar — x, y : floats or arrays is rendered as "y (x,)".
This is probably related to sphinx-doc/sphinx#2227?

Maybe we can the `napoleon_use_param = False` in `conf.py` as suggested. You can try to do that in this PR, or open a separate issue.

> Perhaps, linbasex.basis_cache_cleanup() should be renamed to linbasex.cache_cleanup() for consistency with and dasch.cache_cleanup() basex.cache_cleanup()?

That makes sense to me.

stggh commented on abel/basex.py:
```diff
@@ -129,8 +153,8 @@ def basex_transform(data, nbf='auto', reg=0.0, basis_dir='./', dr=1.0,
     # load the basis sets:
-    Ai = get_bs_basex_cached(n, nbf=nbf, reg=reg,
-                             basis_dir=basis_dir, verbose=verbose)
+    Ai = get_bs_cached(n, sigma=sigma, reg=reg, correction=correction,
```
function name change also within `benchmark.py` line 53: `'basex_bs': basex.get_bs_basex_cached

MikhailRyazanov replied:
Thanks!

MikhailRyazanov commented:
@DanHickstein, `napoleon_use_param = False` seems to work.

@stggh, thanks for noticing the remaining `get_bs_basex_cached()`!
Do you agree with renaming `linbasex.basis_cache_cleanup()` → `linbasex.cache_cleanup()`?

@rth, it seems that `GaussianAnalytical()` had that corrupted docstring [when you first committed it](https://github.com/PyAbel/PyAbel/blob/5da3cd9c77323be017d233a3f1fa82b474ae9f85/abel/analytical.py#L105-L107). Do you know what it was supposed to be?

stggh commented:
> renaming linbasex.basis_cache_cleanup() → linbasex.cache_cleanup()

Yes, feel free to change it in this PR.

MikhailRyazanov commented:
In principle, this could be merged as is, but I would like to hear from @rth (or anybody else) about the `mask_valid` (I can set it to `False` at each end of each interval, if this will help anything) and about
https://github.com/PyAbel/PyAbel/pull/235#discussion_r230621726 and the `GaussianAnalytical` docstring.

Also, what everybody thinks about using `correction=True` by default?

DanHickstein commented:
Yes, if there is any region of the analytic transform where the results are not correct, then set those positions of `mask_valid` to `False`. 

I would imaging that the majority of users would want `correction=True`, so I think that this is an appropriate default. By default we should offer whatever decisions we think produce the best results.

rth commented:
TBH, I don't have much recollection about anything related to `mask_valid`, and looking at the half written docstring I have no idea now what was missing :) If anything doesn't make sense, you might as well re-write it, sorry..

MikhailRyazanov commented:
@rth, something like this?
```
    ratio_valid_sigma : float
        in the benchmark take only the range
        0 < r < ration_valid_sigma * sigma
        (exclude possible artefacts on the axis and the possibly clipped tail)
```

And what about `test_suite="abel.tests.run_cli"` in setup.py?

rth commented:
Yes, that makes sense.

> And what about test_suite="abel.tests.run_cli" in setup.py?

Let's remove it, I'm pretty sure I saw somewhere that specifying `test_suite` in `setup.py` was a historical approach, not very recommended now. It's definitely not useful with pytest.

MikhailRyazanov commented:
@DanHickstein, OK, I'll then set the default `correction=True`.

The mask stuff is not very clear.

One the one hand, if the resulting function is continuous (is there any reason to use discontinuous piecewise polynomials in tests?), then it should be well-behaved everywhere.

One the other hand, if the function has singularities, then the extent of the problematic regions will depend on the transform method more than on the function itself.

Maybe adding a parameter like `valid_gap` to set ±`valid_gap` points around each domain boundary to `False` would be reasonable?

MikhailRyazanov commented:
Final changes
---

BASEX:
* Enabled `correction=True` by default.

Other:
* Renamed `linbasex.basis_cache_cleanup()` to `linbasex.cache_cleanup()` for consistency with other methods (see #232).
* Set `napoleon_use_param = False` in `doc/conf.py` for proper rendering of multiple input parameters with shared type/description. Please check that it did not break something else.
* Corrected the broken `GaussianAnalytical` docstring.
* Removed `test_suite="abel.tests.run_cli"` from `setup.py` (amending #228).

stggh commented:
> which functions had invalid regions

`direct`
   See #214 for issues with the grid/derivative that produced odd end values, now fixed.

`hansenlaw`
   End values are not defined, copied from adjacent columns.

MikhailRyazanov commented:
@stggh, so each method should also have its own `mask`.

Let's leave it as is for now. If some useful thoughts on where and how this can be used appear, then it would be better to make a separate decision, consistent for all functions and methods.

DanHickstein commented:
Is this ready to merge?

MikhailRyazanov commented:
Should be...

stggh commented:
@MikhailRyazanov as the code author, you have the honour of pushing the green button.

MikhailRyazanov commented:
Done.
Thanks everybody for the feedback and testing!

DanHickstein commented:
Awesome! PyAbel has seen some great improvements in the past few weeks :)
